MIGRATION.md

```diff
@@ -0,0 +1,130 @@
+
+## Helper
+
+This release includes v3 of the `algoliasearch-helper` package. If you are using the built-in widgets, nothing will change for you. This version no longer includes `lodash`, having two main places of impact for users. The main difference will be in bundle size, which is now significantly smaller. If you are using any methods from the `helper` or `searchResults`, please see the detailed changelog [here](https://github.com/algolia/algoliasearch-helper-js/blob/next/documentation-src/metalsmith/content/upgrade.md).
+
+## Custom widgets
+
+As mentioned earlier, any reference of the helper now contains the "new helper", as well as the `getConfiguration` life cycle no longer being used. You can replace its usage with `getWidgetSearchParameters` and `getWidgetState`. Note that this means that your custom widget will also take part in routing from then on. You can still exclude it from the URL via `stateMapping`.
+
+### getConfiguration
+
+This life cycle could be used for setting up a facet, or a default refinement previously. This is now replaced by `getWidgetState` and `getWidgetSearchParameters`. This means that there's no longer a separate step for setting up. However, that can be implemented using a check inside `getWidgetSearchParameters`. Code like this (abbreviated):
+
+```js
+const widget = {
+  getConfiguration(searchParams) {
+    return {
+      disjunctiveFacets: ['myAttribute'],
+    };
+  },
+};
+```
+
+Will become like this:
+
+```js
+const widget = {
+  getWidgetSearchParameters(searchParameters, { uiState }) {
+    return searchParameters
+      .addDisjunctiveFacet('myAttribute')
+      .addDisjunctiveFacetRefinement(
+        'myAttribute',
+        uiState.myWidgetName.myAttribute
+      );
+  },
+  getWidgetState(uiState, { searchParameters }) {
+    return {
+      ...uiState,
+      myWidgetName: {
+        myAttribute: searchParameters.getDisjunctiveRefinements('myAttribute')
+      }
+    };
+  }
+};
+```
+
+While this looks longer, more of the functions are now together than before, and it isn't possible anymore to get into a state where `getConfiguration` is called multiple times and isn't behaving as you would want.
+
+## connectAutoComplete
+
+The `indices` option has been removed, in favour of using `index` widgets (see the federated search section). In practice, this:
+
+```js
+const autocomplete = connectAutocomplete(() => {/* ... */});
+
+search.addWidget(
+  autocomplete({
+    indices: [{
+      name: "additional"
+    }]
+  })
+);
+```
+
+Will be replaced with this:
+
+```js
+const autocomplete = connectAutocomplete(() => {/* ... */});
+
+search.addWidgets([
+  index({ indexName: "additional" }),
+  autocomplete()
+]);
+```
+
+## onHistoryChange
+
+This was a function which wasn't documented, and has been a no-op since InstantSearch v2, it has now been removed. Listening to changes in the URL can be done by creating a custom `router` and listening in the `write` hook.
```

Comment:
I think later we'd want to example for this as I don't find it super trivial TBH.

```js
const router = historyRouter()
const originalWrite = router.write.bind(router)
router.write = state => {
    // listening 
    originalWrite(state)
}

{
   routing: {
      router,
      stateMapping: ...
   }
}

```

Reply:
Would add this, but not sure if it’s used